keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+
+#### Release notes
+
+The prow “release-note” plugin must be enabled for the kubeadm repository.
+This will make PRs require release notes or the option NONE. Once a k8s
+release has begun ideally the tool:
+https://github.com/kubernetes/release/tree/master/cmd/release-notes
+should be levereged to obtain all the release notes for kubeadm in this
+release.
+
+The alternative is for the kubeadm maintainers to collect the release
+notes manually and send them to the release team, but this might be more
+difficult to handle as the k8s release tooling sends emails with the release
+notes right away. Also the kubeadm maintainers might make mistakes forgetting
+about a certain feature/bug fix.
+
+### Test Plan
+
+#### CI artifacts
+
+Currently the kubeadm binary is built and uploaded in the CI GCS bucket
+next to the rest of the k8sbinaries like kubelet.
+
+There is a complication as now another repository has to be monitored
+for changes. Currently the CI build happens after a PR merged in the
+k/k repository.
+
+The proposal is to always rebuild kubeadm based on the latest tips of
+the master and release-x.yy branches of the k/kubeadm repository after
+a CI build for the k/k repository triggers. In this case the k/release
+build tooling must not trigger the writing of a KUBE_VERSION file in the
+k/kubeadm repository but rather let the version to generate based on the
+latest Git tags and SHAs.
+
+#### PR testing
+
+There is already support for running pre-submit jobs on k/kubeadm PRs.
+Once kubeadm is moved to the new location a collection of PR blocking
+jobs should be added to make sure the project is always buildable with
+the current version of Go that the k/k repository uses. Syncing Go versions
+should be the responsibility of the k/kubeadm maintainers.
+
+The new PR blocking jobs must support building kubeadm from source with
+the latest commits that a PR for k/kubeadm introduces. The rest of the
+k8s artifacts can be pulled from the CI GCS bucket. This can also allow
+catching problems introduced in k/k when using kubeadm as the deployer.
+
+The /hack/verify-* and update-* scripts that k/kubeadm/kubeadm uses
+will not be a direct copy of k/k but will provide similar results:
+Go versions are in sync Go-fmt, linter and import tools are in sync
+as much as possible
+
+The following lists of tests will be executed on PRs as blocking:
+- Unit tests
+- Integration tests (currently reside in k/k/cmd/kubeadm/test)
+- E2e tests from the e2e-kubeadm suite (currently resides
+in k/k/test/e2e-kubeadm)
+- Full cluster deployment e2e, possibly upgrade and skew tests too
+
+#### Periodics jobs
+
+No changes to the current kubeadm periodic jobs will be required
+as long as the CI build process continues to push the kubeadm
+and the rest of the k8s binaries to the GCS buckets.
+
+### Graduation Criteria
+
+Not applicable as the work in this KEP should be ideally executed within
+one k8s release cycle.
```

Comment:
Since discussion seems to aim to do this during the release 1.18 cycle, should this be explicitly stated?

What is the fallback plan if it slips?

Reply:
Graduation criteria only applies to alpha-beta-GA.
executing cycle is tracked in the k/e issue.

> What is the fallback plan if it slips?

- first fallback is having a KEP exception if we end up with a late consensus.
- if there is no consensus and if the move stalls, the second fallback is to stop releasing kubeadm as part of the k8s artifacts.

this was my primary question to SIG Release when discussions about this started happening.

Reply:
So regardless of the state of this KEP, kubeadm binaries will not come from k/k starting in approximately 9 weeks time, ie: end of March 2020?

Reply:
if the SIG release aspects of this KEP are not accepted we have the option to remove SIG release as a stakeholder and proceed our own way. in the end of the day kubeadm is a deployer (a distribution of k8s) and while we are going to slightly break the users by changing the location of the binary and packages this is something we are willing to do, ideally in the 1.18 timeframe. if too many people vote -1 after the complications we can delay for more time (unknown).

Reply:
after my discussion with @tpepper on zoom we realized that maybe 1.18 is a bit of a stretch. so i'm going to update the KEP with more detail what can happen in 1.18 and 1.19.